gensim/similarities/levenshtein.py

```diff
@@ -0,0 +1,158 @@
+        If you don't care about similarities smaller than a known threshold, a
+        more efficient code path can be taken. For terms that are clearly "too
+        far apart", we will not compute the distance exactly, but we will
+        return zero more quickly, meaning "less than `min_similarity`".
+        Default: always compute similarity exactly, no threshold clipping.
+
+    Returns
+    -------
+    float
+        The Levenshtein similarity between `t1` and `t2`.
+
+    Notes
+    -----
+    This notion of Levenshtein similarity was first defined in section 2.2 of
+    `Delphine Charlet and Geraldine Damnati, "SimBow at SemEval-2017 Task 3:
+    Soft-Cosine Semantic Similarity between Questions for Community Question
+    Answering", 2017 <http://www.aclweb.org/anthology/S/S17/S17-2051.pdf>`__.
+
+    """
+    assert alpha >= 0
+    assert beta >= 0
+
+    max_lengths = max(len(t1), len(t2))
+    if max_lengths == 0.0:
+        return 1.0
+
+    min_similarity = float(max(min(min_similarity, 1.0), 0.0))
+    max_distance = int(floor(max_lengths * (1 - (min_similarity / alpha) ** (1 / beta))))
+    distance = levdist(t1, t2, max_distance)
+    similarity = alpha * (1 - distance * 1.0 / max_lengths)**beta
+    return similarity
+
+
+class LevenshteinSimilarityIndex(TermSimilarityIndex):
+    """
+    Computes Levenshtein similarities between terms and retrieves most similar
+    terms for a given term.
+
+    Notes
+    -----
+    This is a naive implementation that iteratively computes pointwise Levenshtein similarities
+    between individual terms. Using this implementation to compute the similarity of all terms in
+    real-world dictionaries such as the English Wikipedia will take years.
+
+    Parameters
+    ----------
+    dictionary : :class:`~gensim.corpora.dictionary.Dictionary`
+        A dictionary that specifies the considered terms.
+    alpha : float, optional
+        The multiplicative factor alpha defined by Charlet and Damnati (2017).
+    beta : float, optional
+        The exponential factor beta defined by Charlet and Damnati (2017).
+    threshold : float, optional
+        Only terms more similar than `threshold` are considered when retrieving
+        the most similar terms for a given term.
+
+    See Also
+    --------
+    :func:`gensim.similarities.levenshtein.levsim`
+        The Levenshtein similarity.
+    :class:`~gensim.similarities.termsim.SparseTermSimilarityMatrix`
+        Build a term similarity matrix and compute the Soft Cosine Measure.
+
+    """
+    def __init__(self, dictionary, alpha=1.8, beta=5.0, threshold=0.0):
+        self.dictionary = dictionary
+        self.alpha = alpha
+        self.beta = beta
+        self.threshold = threshold
+        super(LevenshteinSimilarityIndex, self).__init__()
+
+    def most_similar(self, t1, topn=10):
+        for _, (t2, similarity) in zip(range(topn), (
+                (t2, similarity) for similarity, t2 in sorted(
```

Comment:
a bit hard to read, can you split this into several code-lines (for readability proposes)

Reply:
I extracted two generators from the loop.